apollo-router/src/configuration/mod.rs

```diff
@@ -416,16 +417,31 @@ impl Configuration {
     pub(crate) fn rust_query_planner_config(
         &self,
     ) -> apollo_federation::query_plan::query_planner::QueryPlannerConfig {
-        apollo_federation::query_plan::query_planner::QueryPlannerConfig {
+        use apollo_federation::query_plan::query_planner::QueryPlanIncrementalDeliveryConfig;
+        use apollo_federation::query_plan::query_planner::QueryPlannerConfig;
+        use apollo_federation::query_plan::query_planner::QueryPlannerDebugConfig;
+
+        let max_evaluated_plans = self
+            .supergraph
+            .query_planning
+            .experimental_plans_limit
+            // Fails if experimental_plans_limit is zero; use our default.
+            .and_then(NonZeroU32::new)
+            .unwrap_or(NonZeroU32::new(10_000).expect("it is not zero"));
```

Comment:
doesn't QueryPlannerDebugConfig already set max_evaluated_plans to `10_000`? 🤔

Reply:
Only if you use `QueryPlannerDebugConfig::default()`. The property is required so we would have to do something like this to propagate it correctly:

```rust
let mut debug = QueryPlannerDebugConfig::default();
if let Some(plans_limit) = configuration.supergraph.query_planning.experimental_plans_limit {
    debug.max_evaluated_plans = NonZeroU32::new(plans_limit).unwrap_or(debug.max_evaluated_plans);
}
```
That's even more ceremony and if we use the `Default` impl, and a new option is added to `QueryPlannerDebugConfig`, we could easily forget to pass it on from the router side.

By doing it like this we duplicate the default value but we are at least forced by the compiler to provide all the options intentionally.

Reply:
makes sense!